oci/oci_image_layout.bzl

```diff
@@ -0,0 +1,99 @@
+    descriptor = ctx.attr.manifest[OCIDescriptor]
+
+    out_dir = ctx.actions.declare_directory(ctx.label.name)
+
+    ctx.actions.run(
+        executable = toolchain.sdk.ocitool,
+        arguments = [
+            "--layout={layout}".format(layout = layout.blob_index.path),
+            "--debug={debug}".format(debug = str(ctx.attr._debug[DebugInfo].debug)),
+            "create-oci-image-layout",
+            # We need to use the directory one level above bazel-out for the
+            # layout-relative directory. This is because the paths in
+            # oci_image_index's index.layout.json are of the form:
+            # "bazel-out/os_arch-fastbuild/bin/...". Unfortunately, bazel
+            # provides no direct way to access this directory, so here we traverse
+            # up 3 levels from the bin directory.
+            "--layout-relative={root}".format(root = ctx.bin_dir.path+"/../../../"),
+            "--desc={desc}".format(desc = descriptor.descriptor_file.path),
+            "--registry={registry}".format(registry = ctx.attr.registry),
+            "--repository={repository}".format(repository = ctx.attr.repository),
+            "--out-dir={out_dir}".format(out_dir = out_dir.path),
+        ],
+        inputs = ctx.files.manifest,
+        outputs = [
+            out_dir,
+        ],
+        use_default_shell_env = True,
+    )
+
+    return DefaultInfo(files = depset([out_dir]))
+
+oci_image_layout = rule(
+    doc = """
+        Writes an OCI Image Index and related blobs to an OCI Image Format
+        directory. See https://github.com/opencontainers/image-spec/blob/main/image-layout.md
+        for the specification of the OCI Image Format directory. Local blobs are
+        used where available, and if a referenced blob is not present, it is
+        fetched from the provided OCI repository and placed in the output.
+
+        In order for this rule to work correctly in its current state, the
+        following flags must be provided to bazel:
+        --incompatible_strict_action_env=false
+        --spawn_strategy=local
+
+        The incompatible_strict_action_env flag is required because in order to
+        access the registry, a credential helper executable (named
+        docker-credential-<SOMETHING>) must be available so that ocitool can
+        execute it. The incompatible_strict_action_env flag makes the system
+        PATH available to bazel rules.
+
+        The spawn_strategy flag must be set to local because currently,
+        oci_image_index is only declaring the new JSON files it creates as
+        outputs; it's not declaring any manifests or layers from the images as
+        outputs. By default, Bazel only permits rules to access specifically
+        declared outputs of the rule's direct dependencies. In order for this
+        rule to access the transitive set of outputs of all dependencies, we
+        must disable bazel's sandboxing by setting spawn_strategy=local.
```

Comment:
Further thoughts:

- The requirement here is to make the entire built image available locally for security scanning
- The built image consists of:
    - the base image (which `oci_pull` currently elides, replacing it with a reference)
    - the layers on top of the base image (which are built locally by `rules_oci`)
- Thus, to create a complete local image, we need a variant of the `oci_pull` repository rule that pulls _all_ the remote layers; this could then be referred to in the provider output of the `oci_image` rule, and used by `oci_image_layout`.

This moves all network access back into repository rules, without too much awkwardness (since you already have an `oci_pull` for the base image).

Reply:
Thank you very much for the thorough review; I really appreciate it. Since I'm doing this to create a test rule that depends on it, I'd very much like the test target to be hermetically executable with `bzl test`. I'd also like to avoid pushing the image to a registry first; the idea being that images should be able to pass this test before we push them.

Your second idea seems very doable to me; create a variant of oci_pull that pulls down all necessary base blobs. The oci_image_layout would then be able to access the output of the oci_pull. It shouldn't be too difficult to make providers in oci_image_layout that access the oci_pull outputs. This seems a very clean way of doing it that matches the way bezel repository rules are intended to be used.

One question that arises is, with this approach, should I create a new rule (e.g. oci_pull_all_local or some similar name), or modify the attributes passed to oci_pull to parameterize it. e.g. have a boolean that describes whether to pull all the blobs locally. I'd be interested to hear your thoughts on this.

Reply:
> One question that arises is, with this approach, should I create a new rule (e.g. oci_pull_all_local or some similar name), or modify the attributes passed to oci_pull to parameterize it. e.g. have a boolean that describes whether to pull all the blobs locally. I'd be interested to hear your thoughts on this.

I think we should have a separate `repository_rule`, `oci_pull_all`.  That should ensure that the expense of downloaded all the layers is only required by the test; building/testing other targets should not trigger the download.  You can DRY things up by having an `oci_pull` macro with a flag that determines whether `oci_pull_all` in addition to the `oci_pull` rule.

Reply:
Circling back on this, if we create an oci_pull_all rule (or have a parameterized macro with a boolean that specifies to do pull_all as well as pull), we would need a way for oci_image (and transitively, oci_image_index) to tell it to pull all. I'm thinking that we would need to add a new (optional) param to oci_image and oci_image index that would be passed down the dependency tree and eventually end up being passed to oci_pull (or used to decide to call oci_pull_all). Does this sound right to you?

It would also mean that any oci_image_index targets that we want to run my new test rule on would need to have this new parameter set. Ideally, the test would somehow pass it in, but other uses of oci_image_index would not do so. Unfortunately, I can't think off the top of my head how to achieve that. Any suggestions welcome :).

If what I've suggested sounds sane to you, I'll look at implementing it this week. Thanks again for your help with this.

Reply:
TL;DR: I think the only place you need an extra parameter is in `oci_image`; everything else can be handled by passing the extra information through via providers.

Could we:
- make oci_image take a new, optional `pulled_base` attr, that would be used to pass in the target created by `oci_pull_all`
- pass that information out in a new provider, say `OCIPulledImageInfo`, that would only be present if `pulled_base` was set 
- oci_image_index (and downstream rules) could then look for that provider on its manifests inputs and pass it on (`OCIPulledImageInfo` would have to be more than just a simple pulled_base to handle the multiple `pulled_base`s on its input; you'd need to consider what information is needed by your new `oci_image_layout` rule, and make it trivially available in the new provider)

Reply:
I thought about it, and that sounds like a good approach. It will mean that oci_image_layout won't work unless all the base images that it transitively depends upon have pulled_base set... but I can't think of any way around that with the way bazel works.

Interestingly, Google's oci_pull (https://github.com/bazel-contrib/rules_oci/blob/main/oci/private/pull.bzl) seems to do want we want for this; it even keeps it all in OCI Image Layout format. They have a todo in there to avoid eager download of layers, but the current code downloads everything. I wonder if we could reuse it. We can discuss that in the meeting I've organized.